### ISSUE TITLE
Is d.ToMsg >> unbox correct?

### ISSUE THREAD

**Issue by bender2k14** [elmish/Elmish.WPF#116]
Consider this line:
https://github.com/elmish/Elmish.WPF/blob/5df8514ce4452883a1e417c9f6f479f545339f9e/src/Elmish.WPF/Binding.fs#L165

Of course the code compiles (since it is in `master`), but I am very surprised by the use of `unbox` there.  I don't think the types make sense.  I was expecting to see `box` so that the whole line would be
```
ToMsg = d.ToMsg >> box
```

Can someone double check if this line is correct?

**Comment by bender2k14** [comment]
> Perhaps try with a value-typed/struct message?

Yes, I think this is a necessary condition to exhibit a bug since I think `box` and `unbox` are no-ops for references types.

> Could you test both alternatives? If box works (and you're sure you've hit that code path), feel free to make a PR.

Definitely.

**Comment by bender2k14** [comment]
I understand the (un)boxing so much better now.

To start, the reason for the (un)boxing is that we want to expose fully/strongly typed binding methods to the user, but we can't* keep all the type parameters around.  When we throw away some of that information by converting from some type parameter to `obj`, we have treat value types and references types differently.

*At least, I think we can't.  The main issue I see is that `BindingData<'model, 'msg>` is recursive.  Specifically, it is recursive via the cases `SubModelData`, `SubModelWinData`, and `SubModelSeqData` (but not via `SubModelSelectedItemData`, so maybe its name should not start with `SubModel`?).  Anyway, the most troublesome type parameters that I see are `'bindingModel`, and `'bindingMsg` because they are the ones uses in the recursively.  If we try to hold onto them, then either we are just delaying the inevitable (i.e. delaying the mapping to `obj` for the type parameters of the children to the type parameters of the grandchildren), or we run into a situation that I actually think is impossible (which is only being able to hold onto a finite number of type parameters but having types for everything in a heterogeneous tree of arbitrary depth).

With that out of the way, here are some improvements that I see related to (un)boxing.

First, we might be able to hold onto more type parameters.  Holding onto more type parameters means that we can remove some calls to `box` and `unbox`.  This would improve performance slightly, but of course the primary advantage would be stronger types within `Elmish.WPF` to help ensure correctness.

Second, I see the implementations of the binding methods as doing two things.  One thing is converting from the inputs from a given simple form to a more complex form that could have been given in some overload.  The other thing is some (un)boxing to discard some type information.  In hindsight, it is not so surprising that these two things are sometimes done together since (un)boxing also converts the inputs to alternative forms.  The easiest way to make an improvement here is to have the simpler overloads call the more complex overloads and then only have the ones of maximal complexity do (un)boxing.  This improvement should be made after the one above since stronger types could change some aspects of this improvements.

**Comment by cmeeren** [comment]
> To start, the reason for the (un)boxing is that we want to expose fully/strongly typed binding methods to the user, but we can't* keep all the type parameters around.

Absolutely right. 👍 

> When we throw away some of that information by converting from some type parameter to `obj`, we have treat value types and references types differently.

Not sure what the relevance of this remark is. In the next paragraph you correctly state the reason for using a weakly typed API (recursive bindings). While boxing has implications for ref vs. value types, AFAIK that's not why we do it.


> The main issue I see is that `BindingData<'model, 'msg>` is recursive.

That's certainly one reason, though IIRC (I may be mistaken) it was also somewhat weakly typed before I took over this project with a rewrite that added support for recursive bindings.

> Specifically, it is recursive via the cases `SubModelData`, `SubModelWinData`, and `SubModelSeqData` (but not via `SubModelSelectedItemData`, so maybe its name should not start with `SubModel`?).

`subModelSelectedItem` is a reasonable name, because while it's not in itself a sub-model binding, it's a binding to the selected item of another `subModelSeq` binding. It's a "workaround" when you can't bind to simpler stuff like the key of a selected item. See #77 for the discussion.

> Anyway, the most troublesome type parameters that I see are `'bindingModel`, and `'bindingMsg` because they are the ones uses in the recursively. If we try to hold onto them, then either we are just delaying the inevitable (i.e. delaying the mapping to `obj` for the type parameters of the children to the type parameters of the grandchildren), or we run into a situation that I actually think is impossible (which is only being able to hold onto a finite number of type parameters but having types for everything in a heterogeneous tree of arbitrary depth).

Yep, AFAIK that's not possible.

> First, we might be able to hold onto more type parameters. Holding onto more type parameters means that we can remove some calls to `box` and `unbox`. This would improve performance slightly, but of course the primary advantage would be stronger types within `Elmish.WPF` to help ensure correctness.

If you find a way to accomplish that, go ahead! Note that it may be a non-trivial tradeoff in terms of complexity in other parts of the implementation, so it may or may not be worth it.

> The easiest way to make an improvement here is to have the simpler overloads call the more complex overloads and then only have the ones of maximal complexity do (un)boxing.

I usually do that, but apparently not with this library. I can't remember if there was a reason. In any case, as long as it preserves functionality, this sounds great!

By the way, before refactoring `Binding` stuff: There are a few recent addition to this file in https://github.com/elmish/Elmish.WPF/commit/f81c4ea1f3fb1cda7b4239c4298db66bb133f635 that I haven't yet added tests for. If you plan to refactor stuff here, it would be great if you could first add (passing) tests for these functions (you can simply copy and adapt tests for similar bindings).

**Comment by bender2k14** [comment]
> > When we throw away some of that information by converting from some type parameter to obj, we have treat value types and references types differently.
> 
> Not sure what the relevance of this remark is. In the next paragraph you correctly state the reason for using a weakly typed API (recursive bindings). While boxing has implications for ref vs. value types, AFAIK that's not why we do it.

I am not saying anything deep here.  Maybe I am speaking from my background in C# where (un)boxing is almost always implicit.

In case it helps, I will repeat myself briefly.  We want to throw away some type parameters, which is some function of type `'a -> obj`, but there is essentially only one correct function of that type.  It is called `box` and has an inverse called `unbox`.

> `subModelSelectedItem` is a reasonable name...

FYI: I agree.  I just wanted to spend a moment of consideration to renaming given the small inconsistency I found.

> By the way, before refactoring `Binding` stuff: There are a few recent addition to this file in f81c4ea that I haven't yet added tests for. If you plan to refactor stuff here, it would be great if you could first add (passing) tests for these functions (you can simply copy and adapt tests for similar bindings).

Will do :)

**Comment by bender2k14** [comment]
Recall that this issue is trying to answer the following question.

Is https://github.com/elmish/Elmish.WPF/blob/9d1e13ce9e159c9147b0ca9aaade9971045bc614/src/Elmish.WPF/Binding.fs#L184 (and [duplicated](https://github.com/elmish/Elmish.WPF/blob/9d1e13ce9e159c9147b0ca9aaade9971045bc614/src/Elmish.WPF/Binding.fs#L190) [twice](https://github.com/elmish/Elmish.WPF/blob/9d1e13ce9e159c9147b0ca9aaade9971045bc614/src/Elmish.WPF/Binding.fs#L199)) correct?

I expected to see
```
ToMsg = d.ToMsg >> box
```

The short answer is that both expressions are equivalent.

Now for the long anwer.  First, the runtime analysis.

> Could you test both alternatives? If box works (and you're sure you've hit that code path), feel free to make a PR.
> ...
> Not sure it matters in practice; AFAIK it works fine as it is. But box may be more correct regarding the intention.

Indeed, you were right.  I tested both `box` vs `unbox` and reference vs value (i.e. all four cases) using the SubModel sample by adding the `Struct` attribute to `Counter.Msg`.  This hits that code path as can be seen by eta expanding and placing a breakpoint.  The behavior was correct / as expected / the same in all four cases.

Now for the static analysis.

[Casting](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/casting-and-conversions-%5Bfsharp%5D#casting-object-types) comes in two forms: [upcating](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/casting-and-conversions-%5Bfsharp%5D#upcasting) and [downcasting](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/casting-and-conversions-%5Bfsharp%5D#downcasting).  I always get confused by those names.  I prefer calling them safe casting and unsafe casting (respectively).  Let's briefly define these two types of casting.
1. Upcast / safe cast: Performed explicitly by the `:>` operator.  It is a static cast, which means that the success of the cast is determined at compile time.
2. Downcast / unsafe cast: Performed explicitly by the `:?>` operator.  It is a dynamic cast, which means that the success of the cast is determined at run time.

Anyway, F# typically requires safe casts to be explicit (unlike in C# where implicit safe casting is commonplace).  However, there is one exception in which F# performs implicit safe casting.  Quoting the documentation for [upcasting](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/casting-and-conversions-%5Bfsharp%5D#upcasting).
> Upcasting is applied automatically when you pass arguments to methods on an object type.

In the case of value types, this upcasting has to come with boxing as well.  As Scott Wlaschin [says](https://fsharpforfunandprofit.com/posts/cli-types/#boxing-and-unboxing)
> If we pass in an `int` [into a function that takes a parameter of type `Object`], it is silently boxed into an object...

Let's repeat the code in question with all type annotations included.
```
// d.ToMsg : obj -> 'a
// unbox<'b> : obj -> 'b
ToMsg : obj -> obj = d.ToMsg >> unbox<obj>
```

The output of `d.ToMsg`, which is of type `'a`, is passed into `unbox<obj>`, which expects something of type `obj`.  This is exactly the situation described by Microsoft's and Scott's documentation.  Thus, our `'a` is implicitly boxed and then passed into `unbox<obj>`.  In general, the return type of `unbox<'a>` is `'a`.  In our case, the return type of `unbox<obj>` is `obj`.  The point of `unbox<>` is to unbox value types while performing an unsafe cast.  So if the type parameter of `unbox<>` is a reference type, then no unboxing occurs.  Furthermore, if that type parameter is `obj`, then no downcasting occurs either.  Therefore, `unbox<obj>` is a fancy way of writing `id<obj>`.

In conclusion, the following are equivalent:
1. `let x : obj -> obj = d.ToMsg >> unbox`
2. `let x : obj -> obj = d.ToMsg >> unbox<obj>`
3. `let x : obj -> obj = d.ToMsg >> box<'a> >> id<obj>`
4. `let x : obj -> obj = d.ToMsg >> box<'a>`
5. `let x : obj -> obj = d.ToMsg >> box`

**Comment by cmeeren** [comment]
Thanks for your extensive analysis! Fixed by merging your PR.
